### PR TITLE
SBR downing when a node is quarantined from the other side

### PR DESCRIPTION
**Pull request by johanandren**
Refs #29565
Continues on top of @patriknw s work in #25961

I don't see the issues mentioned in #24764 but maybe that was with the classic remoting only?

### CODE REVIEW

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:98]: this is using the default timeout, which I guess is 3 seconds here. Maybe good to use an explicit timeout to clarify that it is downing itself because of the ThisActorSystemQuarantinedEvent and not by the ordinary SBR (15 seconds)

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:114]: is this comment obsolete now, if the test works?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:119]: maybe we don't have to wait that long

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:115]: I think we need to look into if there could be some "normal" shutdown scenario when it would send the Artery `Quarantined` message from a node that is not trustworthy.

**Review by patriknw — COMMENTED** [review]
looking good, started repeat job at https://jenkins.akka.io:8498/job/legacy-akka-multi-node-repeat4/

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:115]: The reason this works with Artery, but it didn't with classic, is that we have a `harmless` flag to `quarantine`.

Log of the harmless quarantine:
```
[info] [JVM-2] [INFO] [11/02/2020 12:55:14.004] [DowningWhenOtherHasQuarantinedThisActorSystem-akka.actor.internal-dispatcher-25] [Association(akka://DowningWhenOtherHasQuarantinedThisActorSystem)] Association to [akka://DowningWhenOtherHasQuarantinedThisActorSystem@localhost:53089] having UID [-663280180320468705] has been stopped. All messages to this UID will be delivered to dead letters. Reason: Cluster member removed, previous status [Up]
```
and then it doesn't send the `Quarantined` message at that point

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:111]: I wonder if a scenario where cluster is combined with remoting could be problematic. Should SBR check if the remote node signalling quarantine is actually a apart of the cluster before killing itself?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala]: to stress this some more we can change to:
```
      runOn(first) {
        val sel = system.actorSelection(RootActorPath(third) / "user")
        (1 to 25).foreach { _ =>
          sel.tell(Identify(None), ActorRef.noSender) // try to ping third
          expectNoMessage(200.millis) // no ThisActorSystemQuarantinedEvent
        }
      }
```

seems to work fine

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:111]: I think that could be a good precaution.

Thinking about how a normal cluster with "invalid" sbr decisions would behave in that case.
Let's say 4 nodes. Split 2 / 2 and both sides decides to remain for some (invalid) reason. Like auto-downing. Would they then send `Quarantined` to each other if the partition heals and they try to talk to each other?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala]: added

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:115]: Added a check that the node is part of the cluster (not checking origin member status though, thought about that for a while but couldn't figure out if I should). 

I'll try to cause such a split brain to see what happens.

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: @patriknw I've created a multiJVM test to look at logs and see what happens in such a scenario but I cannot make sense of it, it looks like that even with a test transport blackhole between the two partitions the quarantined message reaches the other side. I can't make sense of it, do you think you could have a look when you have time? The test is in `akka.cluster.SplitBrainQuarantineSpec`.

Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: ok, will take a look (during the week)

Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: @johanandren I see what you mean. See `def outboundControl`
```
.via(outboundTestFlow(outboundContext))
.viaMat(new OutboundControlJunction(outboundContext, outboundEnvelopePool))(Keep.right)
```

That means that the control messages, such as `Quarantined` are injected into the stream after the outboundTestFlow, which handles the outgoing blackhole.

I don't know if we had some reason for that, but you can try to switch order of those two.

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: Aha, thanks for figuring that one out!

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: After changing the `outboundControl`, I expected split brain/blackhole a 4 node cluster into 2, each downing the other side and then reconnecting after a while would lead to both sides shutting down because they are quarantined by the other side but this did not work.

Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: I'll take a look

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:112]: Note that I never pushed that change

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:46]: wrong extends, should be extends DowningWhenOtherHasQuarantinedThisActorSystemSpec

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:49]: DowningWhenOtherHasQuarantinedThisActorSystemSpec

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:51]: DowningWhenOtherHasQuarantinedThisActorSystemSpec

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:46]: Ah, yes, I fixed that friday but never pushed.

Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala]: Now I've pushed that but not the testFlow/ControlJunction re-order.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:102]: second is currently not shutdown by ThisActorSystemQuarantinedEvent. SBR is downing itself. We should use even longer stable-after
```
nodeConfig(second)(ConfigFactory.parseString("akka.cluster.split-brain-resolver.stable-after = 30s"))
```

Looks like `Quarantined` isn't sent. Looking into that.

Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:102]: Quarantine isn't sent because it's only heartbeat messages here, and we have
https://github.com/akka/akka/blob/510e7374d5851f342e1135f080a48a61ea01e716/akka-remote/src/main/scala/akka/remote/artery/InboundQuarantineCheck.scala#L47-L48

It works when changing the test to 
```
      runOn(second) {
        within(10.seconds) {
          awaitAssert {
            // try to ping third (Cluster Heartbeat messages will not trigger the Quarantine message)
            system.actorSelection(RootActorPath(first) / "user").tell(Identify(None), ActorRef.noSender)
            // shutting down itself triggered by ThisActorSystemQuarantinedEvent
            cluster.isTerminated should ===(true)
          }
        }
      }
```

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala:112]: I looked at the logs for this one also. It must send some real messages to be relevant.
```
runOn(first, second) {
        system.actorSelection(RootActorPath(third) / "user").tell(Identify(None), ActorRef.noSender)
        system.actorSelection(RootActorPath(fourth) / "user").tell(Identify(None), ActorRef.noSender)
      }
      runOn(third, fourth) {
        system.actorSelection(RootActorPath(first) / "user").tell(Identify(None), ActorRef.noSender)
        system.actorSelection(RootActorPath(second) / "user").tell(Identify(None), ActorRef.noSender)
      }
      Thread.sleep(5000)
      runOn(first) {
        system.actorSelection(RootActorPath(third) / "user").tell(Identify(None), ActorRef.noSender)
        system.actorSelection(RootActorPath(fourth) / "user").tell(Identify(None), ActorRef.noSender)
      }
      runOn(third, fourth) {
        system.actorSelection(RootActorPath(first) / "user").tell(Identify(None), ActorRef.noSender)
        system.actorSelection(RootActorPath(second) / "user").tell(Identify(None), ActorRef.noSender)
      }
      Thread.sleep(5000)
      enterBarrier("after-pass-through")
      cluster.isTerminated should ===(false)
```

Note that it should not trigger shutdown of other side.

This is what happens:
```
[JVM-1] Quarantine piercing attempt with message [ActorSelectionMessage(akka.actor.Identify)] to [Actor[akka://SplitBrainQuarantine@third/]]

[JVM-1] Dropping message [akka.remote.artery.Quarantined] from [akka://SplitBrainQuarantine@third#-4845019687174480295] because the system is quarantined
```

That is good. It doesn't trigger the ThisActorSystemQuarantinedEvent because the Quarantined is dropped before that. If it would have published ThisActorSystemQuarantinedEvent then SBR would anyway not act on that because it's not from a member.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:115]: I tested with the changed order of outboundTestFlow and seems good, and makes more sense. I think we should make that change.

**Comment by patriknw** [comment]
With some cleanup of the tests I think this is good. We'll run it in a multi-node repeat job before merging.

**Comment by johanandren** [comment]
Alright, thanks for looking through it, I'd never have figured that dont-connect-on-heartbeat!

I can do the cleanup later today.

I didn't really intend to include `SplitBrainQuarantineSpec` in the PR, it was more to manually verify what happens in that scenario, do you think I should I should rather make that an actual test with no 30s sleeps etc and include or is it fine to remove it completely?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/DowningWhenOtherHasQuarantinedThisActorSystemSpec.scala:54]: here is one more SplitBrainQuarantineSpec

**Comment by patriknw** [comment]
I think it can be good to have that SplitBrainQuarantineSpec to verify that they are not shutting down each other after healing, since we said that we don't trust ThisActorSystemQuarantinedEvent when it's from non-member.  Now it doesn't trigger the SBR member check but good anyway. WIthout too much sleeps, yes.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala
Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala:112]: Ok, good, I guess I had bad luck when thinking about what should happen once the partition goes away and thought about it from the aspect that it is an old member downed that turns out to be a zombie, but a quarantined message from a random node not in the cluster is more what it is.

**Comment by johanandren** [comment]
Alright, this is already running on https://jenkins.akka.io:8498/job/legacy-akka-multi-node-repeat4/

So let's see it repeat green for a while. 🤞

File: akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala
Comment (octonato) [inline akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala:330]: The name of this method is misleading as the goal is not to reverse a decision. The exception message is also misleading in that sense. 

We only call that method when we fail to acquire the lease, so we need to convert the `decision` to something that makes sense for the side that didn't get the lease. 

When we reach that point, the only possible decisions are `AcquireLeaseAndDownUnreachable` and `AcquireLeaseAndDownIndirectlyConnected` because that was already narrow down to `AcquireLeaseDecision`.

This could be refactored to: 

```scala
def reverseDecision(decision: AcquireLeaseDecision): Decision = {
  decision match {
    case AcquireLeaseAndDownUnreachable(_)         => DownReachable
    case AcquireLeaseAndDownIndirectlyConnected(_) => ReverseDownIndirectlyConnected
  }
}
```

**Review by octonato — COMMENTED** [review]
I did a first pass. Looking good. 

Check comment on the `reverseDecision`. 

Tomorrow I will review the specs.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala
Comment (octonato) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala:121]:
```suggestion
      runOn(first, second) {
```

To clarify, why are we sending `Identity` messages twice? 
Sending once from each node to each other node on the other side would be enough to get them all quarantined, no?

If we want/need to send twice, we probably want to include `second` to the party. 

Could you add a comment explaining why?

**Comment by johanandren** [comment]
Dang, I thought this was running a repeat job but the jenkins job was disabled. Enabled it now.

File: akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala
Comment (johanandren) [reply akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala:330]: I think it says exactly what it does, it reverses a decision, `DownIndirectlyConnected` becomes `ReverseDownIndirectlyConnected`. I'd prefer to leave this as is and keep the PR to be about one isolated thing, please open up a separate PR if you really think it needs renaming/cleanup.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala
Comment (johanandren) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/SplitBrainQuarantineSpec.scala:121]: It sends Identify to both nodes that was previously on the other side of the partition. It should send _from_ both nodes though, good catch.

File: akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala
Comment (octonato) [reply akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala:330]: Ok, we can do the clean-up on another PR.

Comment (octonato) [reply akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala:330]: I have the code ready and was planning to send a PR on top of this PR, but this PR is from a fork. So, I will wait for this to be merged.

**Comment by johanandren** [comment]
4 days of hourly runs without a failure, repeat job now moved on to repeat something else.

**Comment by johanandren** [comment]
I'll set this up again and let it run for a longer while.

**Comment by johanandren** [comment]
Saw one test failure where down from one side of the partition reached the other, which is very surprising given that there is a barrier after the blackhole, before downing nodes to cause the partition.
I don't think this indicates a problem with the down when quarantined logic though.

https://jenkins.akka.io:8498/job/(legacy)%20akka-multi-node-repeat-johan/623/consoleFull

**Comment by johanandren** [comment]
Saw another failure where the nodes has made a blackhole partition and then down the other side, but after that gossip still reaches across the gossip somehow:

```
[info] [JVM-2] [WARN] [12/23/2020 18:03:43.653] [SplitBrainQuarantine-akka.actor.internal-dispatcher-22] [Cluster(akka://SplitBrainQuarantine)] Cluster Node [akka://SplitBrainQuarantine@localhost:43853] - Received gossip where this member has been downed, from [akka://SplitBrainQuarantine@localhost:45061]
```

Where 45061 is JVM-3. Not sure what is going on here.

https://jenkins.akka.io:8498/job/(legacy)%20akka-multi-node-repeat-johan/698/consoleFull

**Comment by johanandren** [comment]
Saw this debug log entry I added in a failed repeat build, so assumption that gossip sometimes leaks through blackhole with artery is verified, so should be ok to merge this and figure out a solution for that later:

```
[info] [JVM-2] [DEBUG] [01/11/2021 20:29:20.090] [SplitBrainQuarantine-akka.remote.default-remote-dispatcher-10] [InboundTestStage$$anon$2(akka://SplitBrainQuarantine)] inbound message [ActorSelectionMessage(akka.cluster.GossipEnvelope)] before handshake completed, cannot check if remote is blackholed, letting through
```

**Comment by patriknw** [comment]
good, let me take another look at the changes before merging

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
Saw some more fails like described above, created #29963 to track improving the testkit.

**Comment by johanandren** [comment]
Killing the repeat job now, have seen one or two more failures because of #29963 but green except for that.

**Review by chbatey — APPROVED** [review]

**Pull request merged**
